Azure Datasource: WIP - Using data proxy timeout as Prefer header

dylan-asos opened the pull request:
**What this PR does / why we need it**:

Adds a Prefer header to Azure Monitor data source HTTP requests, which allows passing the Data Proxy timeout setting to the Azure Monitor API.

For example, the default data proxy time of '30 seconds' will be passed as 

Prefer: wait=30

which will reduce the query timeout [from 3 minutes to 30 seconds](https://dev.loganalytics.io/documentation/Using-the-API/Timeouts).


**Which issue(s) this PR fixes**:

Fixes #28199

**Special notes for your reviewer**:

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=28211) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=28211) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=28211) it.</sub>

aknuds1 commented:
Does it make sense to include tests for this?

dylan-asos committed:
Using data proxy timeout as Prefer header on Azure Monitor requests

Using data proxy timeout as Prefer header on Azure Monitor requests

Using data proxy timeout as Prefer header on Azure Monitor requests


Applied gofmt to fix lint issue

kylebrandt commented on pkg/tsdb/azuremonitor/azuremonitor-datasource.go:
```diff
@@ -248,6 +248,7 @@ func (e *AzureMonitorDatasource) createRequest(ctx context.Context, dsInfo *mode
 
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Set("User-Agent", fmt.Sprintf("Grafana/%s", setting.BuildVersion))
+	req.Header.Set("Prefer", fmt.Sprintf("wait=%d", setting.DataProxyTimeout))
```
Are you sure that this applies to the azure monitor metrics service / api? I do see documented for the analytics and insights. So perhaps it should be removed here?

Also this should perhaps also be added to insights-analytics-datasource.go ?

dylan-asos replied:
I can only find it documented for analytics and insights - the metrics API would be less prone to variations in query times, so yes will remove from here + look at insights-analytics-datasource

kylebrandt reviewed (changes requested):
I wondering if you can maybe be more specific about the behavior change you are after?

Generally, I don't think this will change the behavior. Because:
 1. We send a request, and that request will have a timeout of X
 2. The remote azure system after getting the request will set the same timeout of X

Given that 1 starts before 2, the timeout on the request in Grafana is going to hit before the timeout on the server side. Therefore we are still hitting the deadline exceeded on the client and will not get the gateway timeout from the server. I'm not sure if this will stop the query on Azure side or not (as I am not sure what happens when the deadline is exceeded, rst maybe?) 

So if we really need this depending on what it gets us, we probably need a datasource timeout setting, and maybe warn if that is equal or less than the datasource proxy

kylebrandt commented:
@dylan-asos Thank you for the additional info, in particular the query audition information. So I think it is a good change, but the  ... sort of race condition? Still needs to be resolved. Basically, the server side timeout needs to be some amount less than the client timeout, or we will still get `client.Timeout execeed' error message because the client will timeout before the server if they are the same.

dylan-asos commented:
@kylebrandt yep makes sense. keep it simple, wait time = (proxy timeout - 1 sec)?

kylebrandt commented:
> @kylebrandt yep makes sense. keep it simple, wait time = (proxy timeout - 1 sec)?

I think it should be an data source setting because:
 - Do not know what people's latency are from their Grafana instance to azure, 1 second might not be long enough
 - One may wish to have different data sources with timeouts for different cases. For example, a group of users that is allowed to make long running queries, and a group that is not (and Enterprise has data source permissions).

For the setting, I think we should:
 - Make it optional configuration, if not defined default the header to dataproxy timeout-1s. (edge case, do not set header if that would make the timeout 0).
 - Warn in the datasource settings save validation if it is equal to or greater than the data proxy timeout
 - If at runtime it is equal to or greater than, override it and log a warning (to handle the case where someone may change the system setting in the future, without being aware of the data source setting)

dylan-asos commented:
@kylebrandt all sounds good, but becomes a bigger piece of work than this PR in it's current state - will take a little while.  

Will mark this as WIP.

cheekp commented:
Is it now possible to configure the Prefer wait for Log Analytics API calls some how. The data proxy timeout does not seem to have an effect in v8.2.5. At a 30 seconds timeout we only see 499 errors, so the root cause or extent of the query delay against the API is not known.
